src/axom/core/Map.hpp

```diff
@@ -0,0 +1,834 @@
+
+        axom_map::Pair<Key, T> ret(&(m_list[ind]), true);
+        return ret;
+      }
+    }
+    axom_map::Pair<Key, T> ret(&(m_end), false);
+    return ret;
+  }
+
+  /*!
+     * \brief Removes item with the given key from the linked list, if item exists. Relinks accordingly.
+     *
+     * \param [in] key the key of the item to be found and removed.
+     *
+     * \return Returns true if item was found and removed, false otherwise. 
+     *  
+     */
+  bool remove(const Key& key)
+  {
+    if(m_head == -1)
+    {
+      return false;
+    }
+    IndexType ind = m_head;
+    IndexType prev = -1;
+    do
+    {
+      if(m_list[ind].key == key)
+      {
+        if(prev != -1)
+        {
+          m_list[prev].next = m_list[ind].next;
+        }
+        if(ind == m_head)
+        {
+          m_head = m_list[ind].next;
+        }
+        m_list[ind].next = m_free;
+        m_free = ind;
+        m_size--;
+        return true;
+      }
+      prev = ind;
+      ind = m_list[ind].next;
+    } while(ind != -1);
+    return false;
+  }
+
+  /*!
+    * \brief Returns pointer to node containing key-value pair matching provided key, if possible.
+    *
+    * \param [in] key the key of the item to be searched for.
+    *
+    * \return Returns pointer to the requested node if possible, and to special end node otherwise.
+    *
+    */
+  axom_map::Node<Key, T>& find(const Key& key)
+  {
+    if(m_head == -1)
+    {
+      return m_end;
+    }
+    IndexType ind = m_head;
+    do
+    {
+      if(m_list[ind].key == key)
+      {
+        return m_list[ind];
+      }
+
+      ind = m_list[ind].next;
+    } while(ind != -1);
+    return m_end;
+  }
+
+  /*!
+   * \brief Returns maximum number of items that can be stored in this Bucket instance.
+   *
+   * \return Returns maximum number of items that can be stored in this Bucket instance.
+   *
+   */
+  int get_capacity() const { return m_capacity; }
+
+  /*!
+   * \brief Returns current number of items in this Bucket instance.
+   *
+   * \return Returns current number of items in this Bucket instance.
+   */
+  int get_size() const { return m_size; }
+  ///@}
+
+  axom_map::Node<Key, T>* m_list;
+  axom_map::Node<Key, T> m_end;
+  IndexType m_head;
+  IndexType m_free;
+  int m_capacity, m_size;
+};
+
+/// @}
+}  // namespace axom_map
+
+/*!
+ * \class Map
+ *
+ * \brief Provides a hashmap implementation, relying upon chaining to 
+ *  resolve hash collisions. Simple hashmap for now, future work will use 
+ *  RAJA and Umpire to allow use of map within kernels. Resembles 
+ *  unordered_map, which we can't use due to a lack of host-device decoration.
+ *
+ * \tparam Key the type of keys. Must allow equality comparison, must be copyable. 
+ * \tparam T the type of values to hold, must be copyable.
+ * \tparam Hash functor that takes an object of Key type and returns 
+ *  a hashed value of size_t (for now)
+ */
+
+template <typename Key, typename T, typename Hash = std::hash<Key>>
```

Comment:
Looking good @ksuarez1423 

Quick early feedback -- I think we'll want to add a `Policy` (or `ExecPolicy`) parameter so the use can explicitly request the execution policy. E.g. in an `OpenMP` configuration of axom, we might still want a sequential hashmap.

Reply:
Implemented, as well as exactly what I needed -- I had been trying to work out the best way to handle guarding things that needed to be guarded, and this answered that perfectly.